projects/bootloader/test/test_ping.c

```diff
@@ -0,0 +1,268 @@
+
+  uint8_t s_destination_nodes[DGRAM_MAX_DEST_NODES_SIZE] = { 0 };
+  CanDatagramRxConfig rx_config = {
+    .destination_nodes = s_destination_nodes,
+    .data = s_rx_data,
+    .node_id = 0,  // listen to all
+    .rx_cmpl_cb = NULL,
+  };
+  can_datagram_start_listener(&rx_config);
+  TestCanDatagramMessage msg;
+  fifo_pop(&s_rx_fifo, &msg);
+  can_datagram_rx(msg.data, msg.len, msg.is_start_msg);
+  MS_TEST_HELPER_AWAIT_EVENT(e);
+  can_datagram_process_event(&e);
+
+  while (can_datagram_get_status() == DATAGRAM_STATUS_ACTIVE) {
+    // listens from s_rx-fifo
+    TestCanDatagramMessage msg;
+    fifo_pop(&s_rx_fifo, &msg);
+    can_datagram_rx(msg.data, msg.len, msg.is_start_msg);
+    MS_TEST_HELPER_AWAIT_EVENT(e);
+    can_datagram_process_event(&e);
+  }
+  TEST_ASSERT_EQUAL(DATAGRAM_STATUS_RX_COMPLETE, can_datagram_get_status());
+  TEST_ASSERT_EQUAL(s_board_id, rx_config.data[0]);
+}
+
+void test_ping_addressed_to_multiple(void) {
+  TEST_ASSERT_OK(ping_init(s_board_id));
+
+  CanDatagramTxConfig tx_config = {
+    .dgram_type = BOOTLOADER_DATAGRAM_PING_COMMAND,
+    .destination_nodes_len = 5,
+    .destination_nodes = &s_board_id,
+    .data_len = 0,
+    .data = NULL,
+    .tx_cmpl_cb = tx_cmpl_cb,
+  };
+  prv_send_tx_to_fifo(tx_config);
+
+  TEST_ASSERT_OK(bootloader_can_register_debug_handler(prv_rx_to_fifo));
+
+  Event e = { 0 };
+  prv_tx_from_fifo();
+  MS_TEST_HELPER_CAN_TX_RX(CAN_DATAGRAM_EVENT_TX, CAN_DATAGRAM_EVENT_RX);
+
+  size_t tx_left = fifo_size(&s_tx_fifo);
+  size_t response_size = 0;
+  bool send_tx = true;
+  while (can_datagram_get_status() == DATAGRAM_STATUS_ACTIVE) {
+    // this loop will loop through both command rx and response tx
+    if (send_tx && tx_left > 0) {
+      prv_tx_from_fifo();
+      send_tx = false;
+      tx_left--;
+    }
+    MS_TEST_HELPER_AWAIT_EVENT(e);
+    if (e.id == CAN_DATAGRAM_EVENT_TX) {
+      send_tx = true;
+    }
+    can_process_event(&e);
+    can_datagram_process_event(&e);
+  }
+  // LOG_DEBUG("rx size: %li\n", fifo_size(&s_rx_fifo));  // should be 3 if rx worked...
+  TEST_ASSERT_EQUAL(DATAGRAM_STATUS_TX_COMPLETE, can_datagram_get_status());
+
+  uint8_t s_destination_nodes[DGRAM_MAX_DEST_NODES_SIZE] = { 0 };
+  CanDatagramRxConfig rx_config = {
+    .destination_nodes = s_destination_nodes,
+    .data = s_rx_data,
+    .node_id = 0,  // listen to all
+    .rx_cmpl_cb = NULL,
+  };
+  can_datagram_start_listener(&rx_config);
+  TestCanDatagramMessage msg;
+  fifo_pop(&s_rx_fifo, &msg);
+  can_datagram_rx(msg.data, msg.len, msg.is_start_msg);
+  MS_TEST_HELPER_AWAIT_EVENT(e);
+  can_datagram_process_event(&e);
+
+  while (can_datagram_get_status() == DATAGRAM_STATUS_ACTIVE) {
+    // listens from s_rx-fifo
+    TestCanDatagramMessage msg;
+    fifo_pop(&s_rx_fifo, &msg);
+    can_datagram_rx(msg.data, msg.len, msg.is_start_msg);
+    MS_TEST_HELPER_AWAIT_EVENT(e);
+    can_datagram_process_event(&e);
+  }
```

Comment:
This stuff (for receiving a full datagram) should also be factored out into a function in the testing library in the next ticket.